opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1610,11 +1610,34 @@ namespace detail {
                 break;
 
             case SURFACE_RATE:
-                for (int phase = 0; phase < np; ++phase) {
-                    if (distr[phase] > 0.0) {
-                        xw.wellRates()[np*w + phase] = target * distr[phase];
+                // assign target value as initial guess for injectors and
+                // single phase producers (orat, grat, wrat)
+                const WellType& well_type = wells().type[w];
+                if (well_type == INJECTOR) {
+                    for (int phase = 0; phase < np; ++phase) {
+                        const double& compi = wells().comp_frac[np * w + phase];
+                        if (compi > 0.0) {
+                            xw.wellRates()[np*w + phase] = target * compi;
+                        }
+                    }
+                } else if (well_type == PRODUCER) {
+
+                    int numProducingPhases = 0;
```

Comment:
Thanks for changing, I think this will be more robust. Just a final nitpick: the numProducingPhases variable name is misleading, since we may have any number of phases producing even with ORAT etc. I guess numPhasesWithTargetsUnderThisControl is a bit long, but that is what it is... Alternatively a comment explaining this (would be a good idea anyway).